System/Random.hs

```diff
@@ -1002,7 +988,39 @@ instance Random Double where
   random = randomDouble
   randomM = uniformR (0, 1)
 
-instance UniformRange Double
+instance UniformRange Double where
+  uniformR (l, h) g = do
+    w64 <- uniformWord64 g
+    let x = word64ToDoubleInUnitInterval w64
+    return $ (h - l) * x + l
+
+-- | Turns a given uniformly distributed 'Word64' value into a uniformly
+-- distributed 'Double' value in the range [0, 1).
+word64ToDoubleInUnitInterval :: Word64 -> Double
+word64ToDoubleInUnitInterval w64 = between1and2 - 1.0
+  where
+    between1and2 = castWord64ToDouble $ (w64 `unsafeShiftR` 12) .|. 0x3ff0000000000000
+{-# INLINE word64ToDoubleInUnitInterval #-}
+
+-- | These are now in 'GHC.Float' but unpatched in some versions so
+-- for now we roll our own.
```

Comment:
Can you add a permalink to gitlab source code where the original comes from?

Reply:
Is this good enough?

https://gitlab.haskell.org/ghc/ghc/-/blob/master/libraries/base/GHC/Float.hs (4bada77d58)

Reply:
:D Gitlab has a button in the top right corner "Permalink", which will give you the link to pagfe that doesn't change: https://gitlab.haskell.org/ghc/ghc/-/blob/6d172e63f3dd3590b0a57371efb8f924f1fcdf05/libraries/base/GHC/Float.hs